internal/backend/gdrive/gdrive.go

```diff
@@ -0,0 +1,566 @@
+
+	contentType := googleapi.ContentType("binary/octet-stream")
+
+	var file *drive.File
+	if existing != nil {
+		file, err = srv.Files.Update(existing.Id, nil).Media(rd, contentType).Context(ctx).Do()
+	} else {
+		file = &drive.File{
+			Name:     name,
+			Parents:  []string{parentID},
+			MimeType: "binary/octet-stream",
+		}
+		file, err = srv.Files.Create(file).Media(rd, contentType).Context(ctx).Do()
+	}
+
+	return file, err
+}
+
+// creates new folder
+func gdriveCreateFolder(ctx context.Context, srv *drive.Service, parentID string, name string) (*drive.File, error) {
+	if name == "" {
+		return nil, errors.New("gdriveCreateFolder: name is required")
+	}
+	dir := &drive.File{
+		Name:     name,
+		MimeType: folderMimeType,
+	}
+	if parentID != "" {
+		dir.Parents = []string{parentID}
+	}
+	dir, err := srv.Files.Create(dir).Context(ctx).Do()
+	if err != nil {
+		return nil, err
+	}
+	return dir, err
+}
+
+func gdriveListFolder(ctx context.Context, srv *drive.Service, parentID string, consumer func(*drive.File) bool) error {
+	q := "trashed = false"
+	if parentID != "" {
+		q += fmt.Sprintf(" and \"%s\" in parents", parentID)
+	}
+	req := srv.Files.List().Q(q).PageSize(folderListPageSize).Fields("nextPageToken, files(id, name)").Context(ctx)
+OUTER:
+	for {
+		resp, err := req.Do()
+		if err != nil {
+			return err
+		}
+		for _, file := range resp.Files {
+			if !consumer(file) {
+				break OUTER
+			}
+		}
+		if resp.NextPageToken == "" {
+			break
+		}
+		req.PageToken(resp.NextPageToken)
+	}
+	return nil
+}
+
+func gdriveDeleteItem(ctx context.Context, srv *drive.Service, itemID string) error {
+	if itemID == "" {
+		return errors.New("gdriveDeleteItem: itemID is required")
+	}
+	return srv.Files.Delete(itemID).Context(ctx).Do()
+}
+
+// TODO should be gdriveNotFound?
+func isNotExist(err error) bool {
+	if er, ok := err.(*googleapi.Error); ok {
+		if er.Code == http.StatusNotFound {
+			return true
+		}
+	}
+	if _, ok := err.(gdriveFileNotFoundError); ok {
+		return true
+	}
+	return false
+}
+
+//
+// gdriveBackend implementation
+//
+
+type gdriveBackend struct {
+	service *drive.Service
+
+	// lazily populated path->id map, guarded by dirsLock
+	dirs     map[string]string
+	dirsLock sync.Mutex
+
+	// used to limit number of concurrent remote requests
+	sem *backend.Semaphore
+
+	backend.Layout
+}
+
+// Ensure that *Backend implements restic.Backend.
+var _ restic.Backend = &gdriveBackend{}
+
+func gdriveFilename(file restic.Handle) string {
+	if file.Type == restic.ConfigFile {
+		return "config"
+	}
+	return file.Name
+}
+
+func gdriveNewService(ctx context.Context, jwtKeyPath string) (*drive.Service, error) {
+	// TODO consider using google's default credentials
+	//      https://developers.google.com/accounts/docs/application-default-credentials
+	if jwtKeyPath == "" {
+		me, err := user.Current()
+		if err != nil {
+			return nil, err
+		}
+		jwtKeyPath = me.HomeDir + "/.config/restic/gdrive-account-key.json"
```

Comment:
* Use a field in the `Config` struct for this
 * The XDG spec (which defines `~/.config`) is more complex than just using a fixed path. If you want to have a look on how it might work, we have code for that in the [cache](https://github.com/restic/restic/blob/5f43e179185ecb96819bfb6301746b5b4be7dec1/internal/cache/dir.go#L15-L26)
 * Don't just concatenate paths by hand, use `filepath.Join()` for that (which is OS dependent and will use the right separator), it'll fail on Windows